pkg/csi/service/wcpguest/controller.go

```diff
@@ -428,30 +444,56 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
 	log.Infof("ControllerUnpublishVolume: called with args %+v", *req)
+
 	err := validateGuestClusterControllerUnpublishVolumeRequest(ctx, req)
 	if err != nil {
 		msg := fmt.Sprintf("Validation for UnpublishVolume Request: %+v has failed. Error: %v", *req, err)
 		log.Error(msg)
 		return nil, err
 	}
 
+	// Retrieve Supervisor PVC
+	svPVC, err := c.supervisorClient.CoreV1().PersistentVolumeClaims(c.supervisorNamespace).Get(ctx, req.VolumeId, metav1.GetOptions{})
+	if err != nil {
+		msg := fmt.Sprintf("failed to retrieve supervisor PVC %q in %q namespace. Error: %+v", req.VolumeId, c.supervisorNamespace, err)
+		log.Error(msg)
+		return nil, status.Error(codes.Internal, msg)
+	}
+	var isFileVolume bool
+	for _, accessMode := range svPVC.Spec.AccessModes {
+		if accessMode == v1.ReadWriteMany || accessMode == v1.ReadOnlyMany {
+			isFileVolume = true
+		}
+	}
+	if isFileVolume {
+		return controllerUnpublishForFileVolume(ctx, req, c)
+	}
+	return controllerUnpublishForBlockVolume(ctx, req, c)
+}
+
+// controllerUnpublishForBlockVolume is helper method to handle ControllerPublishVolume for Block volumes
+func controllerUnpublishForBlockVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest, c *controller) (
+	*csi.ControllerUnpublishVolumeResponse, error) {
+
+	log := logger.GetLogger(ctx)
+
 	// TODO: Investigate if a race condition can exist here between multiple detach calls to the same volume.
 	// 	If yes, implement some locking mechanism
 	virtualMachine := &vmoperatortypes.VirtualMachine{}
 	vmKey := types.NamespacedName{
 		Namespace: c.supervisorNamespace,
 		Name:      req.NodeId,
 	}
+	var err error
 	if err := c.vmOperatorClient.Get(ctx, vmKey, virtualMachine); err != nil {
 		msg := fmt.Sprintf("failed to get VirtualMachines for node: %q. Error: %+v", req.NodeId, err)
 		log.Error(msg)
 		return nil, status.Errorf(codes.Internal, msg)
 	}
 	log.Debugf("Found VirtualMachine for node: %q.", req.NodeId)
-
 	timeoutSeconds := int64(getAttacherTimeoutInMin(ctx) * 60)
 	timeout := time.Now().Add(time.Duration(timeoutSeconds) * time.Second)
-
+	// Block volume support
```

Comment:
nit: You can remove this comment as you have already refactored to have a separate method for block volume support.

Reply:
Done